File: src/main/java/com/papilertus/init/Papilertus.java
package com.papilertus.init;

import com.papilertus.commands.base.FeedbackCommand;
import com.papilertus.commands.core.Command;
import com.papilertus.commands.core.CommandClient;
import com.papilertus.commands.core.CommandListener;
import com.papilertus.gui.button.ButtonListener;
import com.papilertus.gui.modal.ModalListener;
import com.papilertus.listeners.BotJoinListener;
import com.papilertus.plugin.PluginJDA;
import com.papilertus.plugin.PluginLoader;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.JDABuilder;
import net.dv8tion.jda.api.OnlineStatus;
import net.dv8tion.jda.api.requests.GatewayIntent;
import net.dv8tion.jda.api.requests.restaction.CommandListUpdateAction;
import net.dv8tion.jda.internal.requests.Requester;

import javax.security.auth.login.LoginException;
import java.lang.reflect.Field;
import java.lang.reflect.Modifier;

public final class Papilertus {
    private static JDA jda;
    private static PluginJDA pluginJDA;

    public static void main(String[] args) throws LoginException, InterruptedException {
        final Config config = Config.getConfig();
        final PluginLoader loader = new PluginLoader(config.getPluginDir());
        loader.load();

        final CommandClient commandClient = new CommandClient();

        //adding plugin commands
        commandClient.addCommands(loader.getRegisteredCommands().toArray(new Command[0]));

        //adding inbuild commands
        commandClient.addCommands(new FeedbackCommand(Config.getConfig().getFeedbackRecipientId()));

        final JDABuilder builder = JDABuilder.create(config.getToken(), GatewayIntent.getIntents(GatewayIntent.ALL_INTENTS));

        //General listener for the commands
        builder.addEventListeners(new CommandListener(commandClient));

        //General listener for the buttons and modals
        builder.addEventListeners(new ButtonListener());
        builder.addEventListeners(new ModalListener());

        builder.addEventListeners(new BotJoinListener(commandClient));

        //Add Plugin event listeners
        if (loader.getRegisteredEventListeners().size() > 0) {
            for (Object listener : loader.getRegisteredEventListeners()) {
                builder.addEventListeners(listener);
            }
        }

        builder.setStatus(OnlineStatus.ONLINE);

        builder.disableCache(config.getCacheFlags());

        jda = builder.build().awaitReady();

        //Setting fosscord as the api endpoint to provide fosscord compatibility
        if (config.isFosscordMode()) {
            try {
                Field field = Requester.class.getDeclaredField("DISCORD_API_PREFIX");
                field.setAccessible(true);

                Field modifiers = Field.class.getDeclaredField("modifiers");
                modifiers.setAccessible(true);
                modifiers.set(field, field.getModifiers() & ~Modifier.FINAL);
                field.set(null, "https://api.fosscord.com");

            } catch (NoSuchFieldException | IllegalAccessException e) {
                throw new RuntimeException(e);
            }
        }

        //register commands on all guilds
        final CommandListUpdateAction action = jda.updateCommands();
        action.addCommands(commandClient.getData()).queue();
    }

    public static PluginJDA getPluginJDA() {
        if (jda == null)
            return null;
        if (pluginJDA == null) {
            pluginJDA = new PluginJDA(jda);
        }

        return pluginJDA;
    }
}

File: src/main/kotlin/com/papilertus/init/Papilertus.kt
package com.papilertus.init

import com.papilertus.command.CommandClient
import com.papilertus.command.CommandListener
import com.papilertus.command.core.FeedbackCommand
import com.papilertus.command.core.ReloadCommand
import com.papilertus.command.core.UnloadCommand
import com.papilertus.gui.contextMenu.ContextMenuListener
import com.papilertus.gui.contextMenu.core.UserInfoContextMenuEntry
import com.papilertus.gui.modal.ModalListener
import com.papilertus.plugin.PluginLoader
import com.sksamuel.hoplite.ConfigLoaderBuilder
import com.sksamuel.hoplite.addFileSource
import dev.minn.jda.ktx.jdabuilder.createJDA
import kotlinx.coroutines.runBlocking
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.requests.GatewayIntent
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.io.File
import java.nio.file.Files
import kotlin.system.exitProcess

private lateinit var config: Config
lateinit var jda: JDA
    private set
val logger: Logger = LoggerFactory.getLogger("Papilertus")
fun main() {
    val configTomlFile = File("config/Papilertus.toml")

    if (!configTomlFile.parentFile.exists()) {
        configTomlFile.parentFile.mkdir()
    }

    if (!configTomlFile.exists()) {
        logger.info("No config file cannot be found... creating one...")
        val r = ClassLoader.getSystemClassLoader().getResourceAsStream(configTomlFile.name)
        if (r == null) {
            logger.error("An error occurred while creating a logging file")
            exitProcess(1)
        }
        Files.copy(r, configTomlFile.toPath())
        exitProcess(0)
    }

    config = ConfigLoaderBuilder.default()
        .addFileSource(configTomlFile)
        .build()
        .loadConfigOrThrow()

    Database.connect(config)


    jda =
        createJDA(config.token, enableCoroutines = true, intents = GatewayIntent.getIntents(GatewayIntent.ALL_INTENTS))

    val commandClient = CommandClient()

    val loader = PluginLoader(config.pluginDir)
    loader.load()

    Runtime.getRuntime().addShutdownHook(object : Thread() {
        override fun run() = runBlocking {
            logger.info("Shutting down Papilertus and unloading plugins...")
            loader.unloadAll()
            if (Database.connection != null)
                Database.connection!!.close()
        }
    })

    if (!config.disableAllCoreInteractions) {
        commandClient.addCommands(
            FeedbackCommand(config),
            UnloadCommand(loader, commandClient),
            ReloadCommand(loader, commandClient)
        )
        commandClient.addContextMenuEntries(
            UserInfoContextMenuEntry()
        )
    }

    commandClient.addCommands(*loader.getCommands().toTypedArray())
    commandClient.addContextMenuEntries(*loader.getContextMenuEntries().toTypedArray())


    jda.addEventListener(*loader.getListeners().toTypedArray())

    jda.addEventListener(CommandListener(commandClient))
    jda.addEventListener(ContextMenuListener(commandClient))
    jda.addEventListener(ModalListener())

    jda.awaitReady()

    jda.updateCommands().addCommands(commandClient.getData()).complete()
    logger.info("Papilertus started successfully")


}